Revise modelchain.rst for multiple arrays, results attribute (#1223)
pvlib/pvlib-python: docs/sphinx/source/modelchain.rst

    mc.ac

Alternatively, we could have specified single diode or PVWatts related
information in the PVSystem construction. Here we assign parameters for

> PVWatts models to the PVSystem. ModelChain will automatically determine that

Author: I'm not committed to the verb "assign", "pass parameters"?

.. ipython:: python

    mc.run_model(weather);
    mc.results.ac

Of course, these choices can also lead to failure when executing
:py:meth:`~pvlib.modelchain.ModelChain.run_model` if your system objects
do not contain the required parameters for running the model chain.

As a convenience, ModelChain includes two class methods that return a ModelChain
with models selected to be consistent with named PV system models:

* :py:meth:`~pvlib.modelchain.ModelChain.with_pvwatts`
* :py:meth:`~pvlib.modelchain.ModelChain.with_sapm`

> 

Reviewer: worth demonstrating this with one of the methods?

The key parts of ModelChain are:

    1. The ModelChain.run_model methods.
    2. A set of methods that wrap and call the PVSystem methods.
    3. A set of methods that can inspect user-supplied objects to infer

>        the appropriate model when a models isn't specified by the user.

Reviewer: ```suggestion
       the appropriate model when a model isn't specified by the user.
```

run_model methods
~~~~~~~~~~~~~~~~~

ModelChain provides three methods for executing the chain of models. The
methods allow for simulating the output of the PVSystem with different
input irradiance data:

* :py:meth:`~pvlib.modelchain.ModelChain.run_model`, use when ``weather``

>   contains global horizontal, direct and diffuse horizontal irradiance ('ghi', 'dni' and 'dhi')

Reviewer: I'd put ghi, dni, dhi in code blocks with double back ticks because they're required DataFrame column names. Same for poa* and effective_irradiance below.

> assigns default values for temperature (20 C)

Reviewer: maybe worth removing this sentence now so we don't forget when addressing https://github.com/pvlib/pvlib-python/issues/1113

and wind speed (0 m/s) if these inputs are not provided.
:py:meth:`~pvlib.modelchain.ModelChain.prepare_inputs` requires all irradiance
components (GHI, DNI, and DHI). The
:py:meth:`~pvlib.modelchain.ModelChain.complete_irradiance`
method is available for calculating the full set of GHI, DNI, or DHI if
only two of these three series are provided. See also
:ref:`dniestmodels` for methods and functions that can help fully define
the irradiance inputs.

Next, :py:meth:`~pvlib.modelchain.ModelChain.run_model` calls the
wrapper methods for AOI loss, spectral loss, effective irradiance, cell
temperature, DC power, AC power, and other losses. These methods are
assigned to generic names, as described in the next section.

The methods called by :py:meth:`~pvlib.modelchain.ModelChain.run_model`
store their results in the ``results`` attribute, which is an instance of
:py:class:`~.ModelChainResult`. :py:class:`~.ModelChainResult` has the
following attributes:
``weather``, ``times``, ``solar_position``, ``airmass``, ``total_irrad``,
``aoi``, ``aoi_modifier``, ``spectral_modifier``, ``effective_irradiance``,
``cell_temperature``,  ``dc``, ``ac``, ``losses``, ``tracking``,

> ``diode_params``.

Reviewer: Perhaps just include a couple of these and refer the reader to the class documentation for more. I'm worried we're going to forget to update this comprehensive list in future enhancements.

> 

Reviewer 2: The two lines just above here emit deprecation warnings

Inferring models
~~~~~~~~~~~~~~~~

When ModelChain's attributes are not assigned when the instance is created,
ModelChain can infer the appropriate model from data stored on the ``PVSystem``
object. ModelChain uses a set of methods (e.g., :py:meth:`~pvlib.modelchain.ModelChain.infer_dc_model`,
:py:meth:`~pvlib.modelchain.ModelChain.infer_ac_model`, etc.) that examine the
parameters on the user-supplied PVSystem object. The inference methods use set
logic to assign one of the model-specific methods, such as

> :py:meth:`~pvlib.modelchain.ModelChain.sapm` or :py:meth:`~pvlib.modelchain.ModelChain.snlinverter`,

Reviewer 2: ```suggestion
:py:meth:`~pvlib.modelchain.ModelChain.sapm` or :py:meth:`~pvlib.modelchain.ModelChain.sandia_inverter`,
```

to the universal method names ``ModelChain.dc_model`` and ``ModelChain.ac_model``,

> respectively. A few examples are shown below. Inferrence methods generally work

Reviewer 2: ```suggestion
respectively. A few examples are shown below. Inference methods generally work
```

Reviewer: this suggestion might have been buried/overlooked.

    mc.infer_ac_model??

ModelChain for a PVSystem with multiple Arrays
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The PVSystem can represent a PV system with a single array of modules, or

> with [multiple arrays](#PVSystem and Arrays). The same models are applied to

Author: Anyone know how to link to a heading in a different file (pvsystem.rst, in this case)?

                      modules_per_string=10, strings=2)
    array_two = Array(surface_tilt=20, surface_azimuth=160,
                      module_parameters=module_parameters,
                      temperature_model_parameters=temperature_model_parameters,
                      modules_per_string=10, strings=2)
    system_two_arrays = PVSystem(arrays=[array_one, array_two],
                                 inverter_parameters={'pdc0': 8000})
    mc = ModelChain(system_two_arrays, location, aoi_model='no_loss',
                    spectral_model='no_loss')

    mc.run_model(weather)

    mc.results.dc
    mc.results.dc[0]

When ``weather`` is a single DataFrame, these data are broadcast and used
for all arrays. To specify data separately for each array, provide a tuple
for ``weather`` where each element is a DataFrame containing the required data.

Air, module and cell temperatures
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The different run_model methods allow the ModelChain to be run starting with
different irradiance data. Similarly, ModelChain run_model methods can be used
with different temperature data as long as cell temperature can be determined.

> Temperatuer data are passed in the ``weather`` DataFrame and can include:

Reviewer 2: ```suggestion
Temperature data are passed in the ``weather`` DataFrame and can include:
```

* cell temperature (``'cell_temperature'``). If passed in ``weather`` no
  cell temperature model is run.
* module temperature, typically measured on the rear surface (``'module_temperature'``).

>   If found in ``weather`` and ``ModelChain.temperature model='sapm'`` 

Reviewer 2: ```suggestion
  If found in ``weather`` and ``ModelChain.temperature_model='sapm_temp'`` 
```
I think?

  (either set directly or inferred), the :py:meth:`~pvlib.modelchain.ModelChain.sapm_temp`
  method is used to calculate cell temperature.
* ambient air temperature (``'temp_air'``). In this case ``ModelChain.temperature_model``

>   is used to calcualte cell temeprature.

Reviewer 2: ```suggestion
  is used to calculate cell temperature.
```

    from pvlib.pvsystem import Array
    location = Location(latitude=32.2, longitude=-110.9)
    inverter_parameters = {'pdc0': 10000, 'eta_inv_nom': 0.96}
    module_parameters = {'pdc0': 250, 'gamma_pdc': -0.004}
    array_one = Array(surface_tilt=20, surface_azimuth=200,
                      module_parameters=module_parameters,
                      temperature_model_parameters=temperature_model_parameters,
                      modules_per_string=10, strings=2)
    array_two = Array(surface_tilt=20, surface_azimuth=160,
                      module_parameters=module_parameters,
                      temperature_model_parameters=temperature_model_parameters,
                      modules_per_string=10, strings=2)
    system_two_arrays = PVSystem(arrays=[array_one, array_two],
                                 inverter_parameters={'pdc0': 8000})
    mc = ModelChain(system_two_arrays, location, aoi_model='no_loss',
                    spectral_model='no_loss')

    mc.run_model(weather)

    mc.results.dc
    mc.results.dc[0]

When ``weather`` is a single DataFrame, these data are broadcast and used
for all arrays. To specify data separately for each array, provide a tuple

> for ``weather`` where each element is a DataFrame containing the required data.

Reviewer: comment on weather order must be the same as array order or is that too much here?

Author: > this suggestion might have been buried/overlooked.

The text is the same as in the suggestion block, am I missing something?

Reviewer: Not sure why that shows up here. Maybe I did something weird or maybe there is a github bug. Either way, go to line 417 on the files changed tab: https://github.com/pvlib/pvlib-python/pull/1223/files#diff-d8efb4ce3ee848ab1c98f74e88029b4b8ebc2280178abd49cd9895199297c20eR417

Author: Got it. "inferrence" -> "inference"

        mc.results.total_irrad will be a Series (if multiple arrays,
        total_irrad will be a tuple). In this case we put total_irrad
        in a list so that we can iterate. If we didn't put total_irrad
        in a list, iteration will access each value of the Series, one
        at a time.
        The iteration returns a tuple. If there is a single array, pvlib
        unwraps the tuple of length 1. Unwrap the tuple here for consistency
        with the rest of pvlib.
        Returning mc is optional, but enables method chaining.
        """
        if mc.system.num_arrays == 1:
            total_irrads = [mc.results.total_irrad]
        else:
            total_irrads = mc.results.total_irrad
        mc.results.dc = tuple(
            pvusa(total_irrad['poa_global'], mc.results.weather['wind_speed'],
                  mc.results.weather['temp_air'], array.module_parameters['a'],
                  array.module_parameters['b'], array.module_parameters['c'],
                  array.module_parameters['d'])
            for total_irrad, array
            in zip(total_irrads, mc.system.arrays))
        if mc.system.num_arrays == 1:

>             mc.results.dc = mc.results.dc[0]

Reviewer: doesn't ``ModelChainResult.__setattr__`` take care of this for us?

>     def pvusa_ac_mc(mc):

Author: @kanderso-nrel I think it would be worth doing, as an example of how to use the multi-array functionality.

I'm not sure the `if mc.system.num_arrays == 1` condition is needed. PTAL.

>     def pvusa_ac_mc(mc):

Reviewer 2: It is a little clunky but OK with me, especially as I am skeptical that many people use user-defined models anyway.  I guess it could use a comment or two explaining why the complexity is needed.

Here's an untested alternative that I'm not convinced is any cleaner:

```python
def pvusa_mc_wrapper(mc):
    if mc.system.num_arrays == 1:
        total_irrads = [mc.results.total_irrad]
    else:
        total_irrads = mc.results.total_irrad

    modeled_powers = tuple(
        pvusa(total_irrad['poa_global'], mc.results.weather['wind_speed'],
              mc.results.weather['temp_air'], array.module_parameters['a'],
              array.module_parameters['b'], array.module_parameters['c'],
              array.module_parameters['d'])
        for total_irrad, array
        in zip(total_irrads, mc.system.arrays)
    )
    if mc.system.num_arrays == 1:
        mc.results.dc = modeled_powers[0]
    else:
        mc.results.dc = modeled_powers

    return mc
```
